azDevOps/azure/templates/v2/steps/build-node.yml

```diff
@@ -0,0 +1,161 @@
+    parameters:
+      azure_tenant_id: "$(azure_tenant_id)"
+      azure_subscription_id: "$(azure_subscription_id)"
+      azure_client_id: "$(azure_client_id)"
+      azure_client_secret: "$(azure_client_secret)"
+      login_azure: true
+
+  # Audit dependencies
+  - ${{ if eq(parameters.build_audit, true) }}:
+      - template: ./build-audit-dependencies-node.yml
+        parameters:
+          versionSpec: ${{ parameters.versionSpec }}
+          workingDirectory: ${{ parameters.project_root_dir }}
+
+  # Install dependencies
+  - template: ./build-install-dependencies-node.yml
+    parameters:
+      versionSpec: ${{ parameters.versionSpec }}
+      workingDirectory: ${{ parameters.project_root_dir }}
+
+  # Linting and formatting validation
+  - ${{ if eq(parameters.lint_formatting, true) }}:
+      - template: ./test-static-validate-node.yml
+        parameters:
+          versionSpec: ${{ parameters.versionSpec }}
+          workingDirectory: ${{ parameters.project_root_dir }}
+          lintingSharedConfigPackage: ${{ parameters.shared_eslint_config }}
+
+  # Linting and unit tests
+  - ${{ if eq(parameters.unit_test, true) }}:
+      - template: ./test-unit-node.yml
+        parameters:
+          versionSpec: ${{ parameters.versionSpec }}
+          workingDirectory: ${{ parameters.project_root_dir }}
+
+  # Functional tests (e2e)
+  - ${{ if eq(parameters.e2e_test, true) }}:
+      - template: ./test-functional-cypress.yml
+        parameters:
+          env_vars: ${{ parameters.e2e_env_vars}}
+          workingDirectory: ${{ parameters.project_root_dir }}
+
+  # SonarCloud
+  - ${{ if eq(parameters.static_code_analysis, true) }}:
+      - template: ./test-static-code-sonar.yml
+        parameters:
+          workingDirectory: ${{ parameters.project_root_dir }}
+
+  # Consumer driven contract tests
+  - ${{ if eq(parameters.contract_test, true) }}:
+      - template: ./test-contract-pact.yml
+        parameters:
+          versionSpec: ${{ parameters.versionSpec }}
+          build_number: ${{ parameters.docker_imagetag }}
+          workingDirectory: ${{ parameters.project_root_dir }}
+
+  # Check with Pact broker if safe to proceed with deployment (breaking contract changes)
+  - ? ${{ if and(eq(parameters.contract_test, true), eq(parameters.contract_check_can_deploy, true)) }}
+    : - template: ./deploy-contract-pact-can-deploy.yml
+        parameters:
+          workingDirectory: ${{ parameters.project_root_dir }}
+
+  # STEP build NPM for Azure registry
+  - ${{ if eq(parameters.build_on_server, true) }}:
+      - task: Npm@1
+        displayName: Build on ADO using custom feeds
+        inputs:
+          command: "custom" # Options: install, publish, custom
+          customCommand: "ci --only=prod"
+          workingDirectory: ${{ parameters.project_root_dir }}
+          customRegistry: "useNpmrc"
+
+  # Build Docker Image, Scan and Push to Repository
+  - ${{ if eq(parameters.docker_build, true) }}:
```

Comment:
I think it is too big for a step template.
this script could probably be split into two
Build NPM
Build Docker image (Shared with dot-net template)

Reply:
yeah 👍 that is actually a meta step I will remove and add to user generated template so they have control.

still some refactoring to do.